### PR TITLE
Ports: An automation step for generating table with available ports

### PR DESCRIPTION
**Pull request by konradekk**
Added two fields, `description` and `website`, to each `package.sh` script with contents extracted from two columns, “Name” and “Website” resp., in `AvailablePorts.md` table. This enables generating the whole table from the complete data now available in the installer scripts.

The above is done with new `generate-available-ports.py` script. This generator is set to be run as a final step of package development mode (`./package.sh dev`) or on demand. Linter was also slightly updated to accommodate.

### CODE REVIEW

**Comment by GMTA** [comment]
1) We should run updates automatically instead of manually, similar to how `package.sh dev` updates the `README.md` and patches per port for example.
2) Should happen before merge, not after. We can deal with conflicts the usual way (i.e. regenerate after rebase).
3) I don't think we've depended on an external node.js before, we do have Bash and Python scripts. I think it's preferable to stay closer to the build dependencies we already have.

**Comment by konradekk** [comment]
@GMTA , not sure if i get it right; did you mean:
- the idea is okay (i, jelle, like automation), but…
- …run moment is indicated wrongly, and…
- itʼd be better to use bash/python here instead.

to deal with those, (a) i can skip first point, (b) iʼd need to update the description for the pull request, (c) i could go with python, if thatʼs what you hinted at. 😅

**Comment by GMTA** [comment]
@konradekk I very much like the idea of generating `AvailablePorts.md`! It's annoying to have to deal with a large table that effectively consists almost entirely (and thanks to your PR, completely) of properties readily available to us.

Ideally if someone adds or updates a port, the PR they send in already contains the necessary changes to `AvailablePorts.md`. We already do automatic updates of port-related files when running `package.sh dev`, so my suggestion was to include something there to update the available ports as well.

The two scripts you reference are run as part of a CI task so it's not really expected for people to run them locally. Making it a Python script would definitely help keep things a bit more consistent, yeah :^)

**Comment by konradekk** [comment]
@GMTA i think i got your point. two more question, though:
- shall iʼll try to check if i can squeeze a node script run there? (or youʼd like to avoid that altogether?) 
- can i assume that both envvars for python and ruby are in place? if not, where to get them from?
(i regenerate the table as a whole, python and ruby `package.sh` scripts use those…)

we can avoid playing with scripts for now and just update portsʼ definitions here.

**Comment by GMTA** [comment]
@konradekk 

> (or youʼd like to avoid that altogether?)

Ideally, yeah. Having fewer dependencies means less hassle when people want to work on ports. I _think_ we only require things like `coreutils` and `bash` at the moment, with some additional dependencies depending on the port you'd like to build. Expecting people to have a certain node.js version locally goes a bit too far for my taste.

> can i assume that both envvars for python and ruby are in place?

What do you need Ruby for? And the `python` / `ruby` ports' build dependencies are largely irrelevant to generate `AvailablePorts.md`. Would it be possible to convert the code to Bash in `.port_include.sh`, similar to `do_generate_patch_readme`?

**Comment by konradekk** [comment]
> Expecting people to have a certain node.js version locally goes a bit too far for my taste.
as for the node, itʼs about any version having 2-3y, so not necessarily a ‘certainʼ version! 😅

> What do you need Ruby for? And the `python` / `ruby` ports' build dependencies are largely irrelevant when generating `AvailablePorts.md`. 
i did not notice that thereʼre `version.sh` files `source`d in `package.sh` files for those ports. all square here! thanks!

> […] I _think_ we only require things like `coreutils` and `bash` at the moment, […]. Would it be possible to convert the code to Bash in `.port_include.sh`, similar to `do_generate_patch_readme`?
i must confess, i donʼt like bash much—iʼll take a look but somewhat reluctantly here! 😔
(itʼd be glad using python here, is that something youʼd accept instead?)

**Comment by GMTA** [comment]
@konradekk Python is fine, we also use it in our pre-commit hooks IIRC.

**Comment by konradekk** [comment]
@GMTA: please have a look, hopefully itʼs better now! 🤔

File: Ports/zstd/package.sh
Comment (GMTA) [inline Ports/zstd/package.sh:5]: Please use single quotes (`'`) for uninterpreted strings in Bash

File: Ports/generate-available-ports.py
Comment (GMTA) [inline Ports/generate-available-ports.py:1]: I think `generate-available-ports.py` is better suited to live in `/Meta` instead of `/Ports`, since you do not need it to install or run ports.

File: Meta/lint-ports.py
Comment (GMTA) [inline Meta/lint-ports.py:349]: `ordered_list` is not really a descriptive name. `table_keys_sorted`?

File: Ports/.port_include.sh
Comment (GMTA) [inline Ports/.port_include.sh:742]: This can be just `./generate-available-ports.py > AvailablePorts.md` I think?

File: Ports/.port_include.sh
Comment (GMTA) [inline Ports/.port_include.sh:904]: `generate_available_ports` to be consistent with `generate_patch_readme`

File: Ports/README.md
Comment (GMTA) [inline Ports/README.md:295]: I think these `Ports/README.md` changes should not have their own commit, but should be incorporated into the commits with the relevant changes instead.

File: Ports/.port_include.sh
Comment (GMTA) [inline Ports/.port_include.sh:934]: This new command should be added to `Ports/README.md` as well

File: Ports/generate-available-ports.py
Comment (konradekk) [reply Ports/generate-available-ports.py:1]: i was thinking whether put it here or there myself.
- it is used in `Ports` in dev mode—because of that i feel it still belongs there, just as `README.md` generator which just happens to be a part of `./port_include.sh`.
- it seems thereʼs also no call for `./Meta` scripts elsewhere so far—not sure if any place should now about `Meta`…? 🤔

still, i will follow your suggestion here if you believe itʼll be better there; your call? 🙂

File: Ports/.port_include.sh
Comment (konradekk) [reply Ports/.port_include.sh:904]: that was event not an inconsistency but a mistake! 😝

File: Ports/.port_include.sh
Comment (konradekk) [reply Ports/.port_include.sh:742]: definitely!

File: Ports/.port_include.sh
Comment (konradekk) [reply Ports/.port_include.sh:934]: it is mentioned in “dev” section. thereʼs a bunch of other commands not described there, e.g. `generate_patch_readme`…

i guess that youʼd like to have them also there? 🤔
i would give a sentence for each from myself, but i put whatever you provide me with (helping hand highly appreciated!). 🙂

File: Meta/lint-ports.py
Comment (konradekk) [reply Meta/lint-ports.py:349]: hmm… now i think that it should be something like `expected_ports_list` or something; i think i do not have a better name here—iʼll just restore previous version.

File: Ports/zstd/package.sh
Comment (konradekk) [reply Ports/zstd/package.sh:5]: sure thing!

File: Ports/AvailablePorts.md
Comment (ADKaster) [inline Ports/AvailablePorts.md:5]: Hm. The extra padding in this script was intentional so that adding a new Port with a longer name, longer description, or longer website name wouldn't cause a full table reformat commit.

Comment (konradekk) [reply Ports/AvailablePorts.md:5]: i feel that when generating itʼs not so painful; i could provide some minimal lengths that could be updated if needed but not really sure—at the moments itʼs just “snug-in”.

diff also does not look bad there (see the screenshot).
example-diff

the length difference on table columns are: 10 chars for “port”, 2 chars for “name”, none for “version” and 10 chars for “website”. shall i provide this minimal length to match current formatting or is it okay to keep minimal lengths?

File: Ports/README.md
Comment (konradekk) [reply Ports/README.md:295]: should be better now!

File: Ports/AvailablePorts.md
Comment (ADKaster) [reply Ports/AvailablePorts.md:5]: It would be best to pad them out to the current lengths to minimize the change from this patch, and mitigate any full reformats from additional ports in the future. Ideally a "reformat using new script" commit in this PR would only be correcting incorrect information rather than completely redoing the table.

Comment (konradekk) [reply Ports/AvailablePorts.md:5]: iʼll accommodate for some reasonable, round length there. 👌
(shall i keep “version” intact or even avoid changes, or rather provide a new baseline?)

Comment (ADKaster) [reply Ports/AvailablePorts.md:5]: Padding them out to a round number seems reasonable. I can only assume the previous lengths were "enough padding" to be a certain number larger than the (at the time) longest string per column.

Comment (konradekk) [reply Ports/AvailablePorts.md]: kept the previous ones for; have a look. 🙂

File: Ports/generate-available-ports.py
Comment (ADKaster) [inline Ports/generate-available-ports.py:42]: A few comments on the python here:

- we should prefer to use pathlib over os.path <https://docs.python.org/3/library/pathlib.html>
- Can we not do `f'{port}/package.sh showproperty ....'` rather than doing a cd + exec?
- What does calling `sys.stderr.write/flush` from each async task do to our throughput in reading all these files at once?
- I'm not convinced that repeated calls to `sys.<stream>.write` in a row is particularly pythonic. Is there a way we can create strings and then write them all at once? Mostly thinking about the final writeout of the entire file.
- Why are we writing directly to stdout instead of passing a file argument in?

File: Meta/lint-ports.py
Comment (ADKaster) [inline Meta/lint-ports.py:14]: Is the regex change here and the array vs non-array change below part of adding the new script? or just general bugfixes? or leftovers from a different approach?

**Review by ADKaster — COMMENTED** [review]
Generally good with these changes, though I doubt GMTA will get back to the comments before the end of FOSDEM.

File: Meta/lint-ports.py
Comment (konradekk) [reply Meta/lint-ports.py:14]: itʼs needed if you exactly match the length of the content (thatʼs `"<spaces>" in group 3` from the comment above), i.e. `| ××××× |`, so with a single space from the pipe in the “name” column.

treat it as a general fix; was very important for “snug-in” approach but now it can just happen—no need to alert if it everything fits (that could happen if you provide a description with length exceeding of current minimal width without changing the minimum, i.e. just by working with the port).

File: Ports/generate-available-ports.py
Comment (konradekk) [reply Ports/generate-available-ports.py:42]: thanks! answering to each point:
* got rusty on that; did not even it was a thing—definitely it is—will switch to it; thanks!
* it is needed in some form (`package.sh` contains relative call in the prologue), but can be simplified;
* itʼs needed to see the progress; on machine the penalty is negligible (total time sticks around 3.14s).
* i was not aiming to be 100% pythonic here; iʼll make it less offensive by concatenating it. 😉
* it was just easy (not dealing with opening files etc.); shall i take it as a request for change? 🙂

File: Ports/generate-available-ports.py
Comment (GMTA) [reply Ports/generate-available-ports.py:1]: > it is used in Ports in dev mode

Yeah, it is a meta-functionality but we indeed have README generation in there as well. Let's leave it for now :^)

File: Ports/README.md
Comment (GMTA) [inline Ports/README.md:113]:
```suggestion
- After leaving the dev shell, all patches are updated and the user will be prompted
  whether they wish to generate a new patch readme file and update the list of
  available ports.
```

File: Ports/.port_include.sh
Comment (GMTA) [reply Ports/.port_include.sh:936]: Hm, I guess mentioning it under `dev` is enough!

File: Ports/README.md
Comment (konradekk) [reply Ports/README.md:113]: changed!

**Comment by konradekk** [comment]
is there anything else to be addressed or is it good to go?

File: Meta/lint-ports.py
Comment (linusg) [inline Meta/lint-ports.py:76]: `print("...", file=sys.stderr)` everywhere

File: Ports/README.md
Comment (linusg) [inline Ports/README.md:280]: no use of fancy quotes, `'` is fine

File: Ports/README.md
Comment (linusg) [inline Ports/README.md:362]: useless noise, there's hundreds of these across the project - fixing one is not helpful.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Ports/README.md
Comment (konradekk) [reply Ports/README.md:362]: well… 🤷‍♂️

File: Ports/README.md
Comment (konradekk) [reply Ports/README.md:280]: iʼd say just «quotes» 😅. (unicode — your days are still to come! 😆)

File: Meta/lint-ports.py
Comment (konradekk) [reply Meta/lint-ports.py:76]: sure thing!

File: Ports/generate-available-ports.py
Comment (ADKaster) [inline Ports/generate-available-ports.py:120]: Two tiny nitpicks that I won't be too sad if another maintainer merges before they're addressed: 

1) sys.stdout.write --> `print()` to match w/linus's previous comment about `sys.stdout.write(file=sys.stderr)` (iirc stdout is the default for print).

2) `chdir(Path(__file__).parent / '..' / 'Ports')` 

Is this not `chdir(Path(__file__).parent)` ?

Comment (konradekk) [reply Ports/generate-available-ports.py]: ad 1. missed that — nicely caught! 🙇🏻‍♂️
ad 2. it is related to the second bullet in your previous comment—we need to be in that directory no matter the place we are currently, that because of relative references in the `package.sh` prologues (i.e. it is possible to call the script outside of `Ports` directory).

**Review by linusg — COMMENTED** [review]
There's a bunch of print statements where you needlessly add a \n now

**Comment by konradekk** [comment]
> There's a bunch of print statements where you needlessly add a \n now

Not really, all should be good there — those new lines where previously double new lines. 😄

**Review by linusg — CHANGES_REQUESTED** [review]
one more thing: as I think you've been told before, please use a valid email address for your commits. Inventing one on a domain you don't own is bad taste, `@serenityos.org` is available to anyone with 100+ commits. In case you don't want to publish your actual email address for any reason, use the GitHub-provided `<username>@users.noreply.github.com`.

**Comment by konradekk** [comment]
> one more thing: as I think you've been told before, please use a valid email address for your commits. Inventing one on a domain you don't own is bad taste, `@serenityos.org` is available to anyone with 100+ commits. In case you don't want to publish your actual email address for any reason, use the GitHub-provided `<username>@users.noreply.github.com`.

excuse my bad taste! iʼd still attribute my contributions sourcing in serenityos.org and stick to the current state of matters. i believe that misattribution is just temporary as i plan to earn the real e-mail alias/account in the serenityos domain. 🙂

**Comment by GMTA** [comment]
Hi @konradekk , agreed with @linusg that the email address is not appropriate yet. To get to the number of commits needed, you need to get enough PRs merged first :^)

**Comment by AtkinsSJ** [comment]
The way that your commit count is measured, it won't count commits that aren't associated with an email address connected with your GitHub account. So far, that's zero: https://github.com/SerenityOS/serenity/commits/?author=konradekk

You genuinely need to use a proper email if you want it to actually count your work so that you earn the serenityos.org email. :^)

If you want to keep your personal email address private, GitHub assigns one automatically, and you can use that. See here: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address

**Comment by konradekk** [comment]
> The way that your commit count is measured, it won't count commits that aren't associated with an email address connected with your GitHub account. So far, that's zero: https://github.com/SerenityOS/serenity/commits/?author=konradekk
> 
> You genuinely need to use a proper email if you want it to actually count your work so that you earn the serenityos.org email. :^)
> 
> If you want to keep your personal email address private, GitHub assigns one automatically, and you can use that. See here: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address

thatʼs somewhat catch-22 for me here — iʼd like to have a single attribution and i chose to start away (somewhat untastefully) with the target identification. i donʼt believe people would like to rewrite git history for that! 😅

as for tracking the amount of commits, i believe thereʼs still much work on my side in the project, so itʼll take few moments to achieve the commit number. i promise that iʼll count them on my own and notify you immediately after i achieve it. is that achievable?

**Comment by alimpfard** [comment]
> thatʼs somewhat catch-22 for me here — iʼd like to have a single attribution and i chose to start away (somewhat untastefully) with the target identification. i donʼt believe people would like to rewrite git history for that!

Please use a valid email, you can update the mailmap once you get the `@serenityos` email so git treats all your emails as one.
Let's not borrow from the future :)

**Comment by ADKaster** [comment]
@konradekk Given that 5 maintainers have separately asked you to not use the @serenityos.org email on your commits before reaching 100 commits, I wouldn't expect to have any of your PRs merged until that is resolved.

**Comment by konradekk** [comment]
> @konradekk Given that 5 maintainers have separately asked you to not use the @serenityos.org email on your commits before reaching 100 commits, I wouldn't expect to have any of your PRs merged until that is resolved.

oh! got that! 🙇🏻‍♂️

**Comment by linusg** [comment]
why do you keep rebasing this PR without resolving the blocking aspect of it?

**Comment by linusg** [comment]
Closing, feel free to reopen when you decide to change your commit email to something proper. This will not get merged otherwise.

**Comment by konradekk** [comment]
sure thing; thanks! 🙇🏻‍♂️

**Pull request closed without merging**
